keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
+- Explain or define the scenario of multiple KUDO versions with possibly multiple KUDO API versions running on a single Kubernetes cluster
+- Explain or define the planned implementation of operator<->operator dependency graphs (e.g., Kafka operator based on Kafka 2.3.1 requires a Zookeeper operator instance with version above 3.4.10)
+- Explain or define the relationship between KUDO version and KUDO API version
+- Explain or define implementations for transparent and automatic handling of multiple package registries from the KUDO CLI program
+
+## Current State
 
-## Proposal
+An `operator.yaml` file provides versions for all components describing an operator:
 
-A `operator.yaml` file already provides versions for all components describing an operator:
-  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
-  * The bundled application version is provided by the `appVersion` field
-  * The version of the operator is provided by the `version` field
-  * The version of Kubernetes is provided by the `kubernetesVersion` field
+- The _KUDO API version_ is specified in the `apiVersion` field
+- The _KUDO version_ is specified in the `kudoVersion` field
+- The _application version_ is specified in the `appVersion` field
+- The _operator version_ is specified in the `version` field
+- The minimum Kubernetes version is specified in the `kubernetesVersion` field
 
 From these versions, currently only the `version` field can be used when installing packages with `kubectl kudo install`. Furthermore, a package repository index only contains `appVersion` and `version` fields in its entries. This limits the possible installation scenarios.
-At least the following installation scenarios should be supported:
-  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
-  2. A user should be able to install a specific `appVersion` of an operator
 
-### Support multiple API versions in the package repository
+## Proposals
+
+### With regards to package version and operator version
+
+Two proposals: singular and composite package/operator version. In the case of potential operator "bundles" (operators containing multiple app versions) both would end up "hiding" those app versions behind the package/operator version.
+
+#### Singular package/operator version (largely, the current state of things)
+
+##### Pros
+
+- Visually simple
+- It can be known just from the package version whether or not it contains a certain operator-related feature or bug fix
+
+##### Cons
+
+- Operator users care about the application version, not a "synthetic" version
+- Makes it complex to have an "app version"-based backport strategy. Customers might not be able to use operator based on newer app versions due to constraints outside of our control, but would still want to receive new operator-related features and bug fixes
```

Comment:
But users will still be able the install the latest operator revision for a given app version with the changes that are proposed in KEP-19. That's one of the goals of the KEP. Or am I missing something here? Probably relates to what "hiding" means here.